url2a908001cd_malictus_klang_tgz-pJ8-malictus_klang_test_PrimitiveTestJ8/src/malictus/klang/primitives/NullTerminatedString.java

```diff
@@ -0,0 +1,236 @@
+/**
+ * This file is distributed under a BSD-style license. See the included LICENSE.txt file
+ * for more information. 
+ * Copyright (c) 2009, James Halliday
+ * All rights reserved.
+ */
+package malictus.klang.primitives;
+
+import java.io.*;
+import java.nio.*;
+import java.nio.charset.*;
+
+import malictus.klang.KlangConstants;
+
+/**
+ * Primitive representing a null-terminated string.
+ * @author Jim Halliday
+ */
+public class NullTerminatedString implements Primitive, StringPrim {
+	
+	private String value;
+	private String textEncoding;
+	private boolean canChangeEncoding;
+	
+	/**
+	 * Constructor from a file reference. The value will be read from the file at
+	 * the current file pointer, and the file pointer will be advanced to 
+	 * just beyond the string. Note that this is currently the ONLY way to initialize this
+	 * value while specifying a stop point.
+	 * 
+	 * @param raf a RandomAccessFile object that points to a valid file, queued for reading
+	 * @param stoppoint the maximum byte position to stop reading data, or -1 if should read indefinitely
+	 * @param textEncoding the text encoding for this string
+	 * @throws IOException if the file cannot be read properly
+	 * @throws BadValueException if value is incorrect
+	 * @throws CharacterCodingException if a character encoding problem occurs
+	 */
+	public NullTerminatedString(RandomAccessFile raf, long stoppoint, String textEncoding, boolean canChangeEncoding) throws IOException, BadValueException, CharacterCodingException {
+		this.textEncoding = textEncoding;
+		this.canChangeEncoding = canChangeEncoding;
+		if (stoppoint != -1) {
+			if (raf.getFilePointer() >= stoppoint) {
+				value = "";
+				return;
+			}
+		}
+		//first, just find the null byte
+		long curpoint = raf.getFilePointer();
+		
+		byte next = raf.readByte();
+		boolean keepgoing = true;
+		while ( (keepgoing) && (next != 0) && (raf.getFilePointer() < raf.length()) ) {
+			next = raf.readByte();
+			if ( (stoppoint != -1) && (next != 0) ) {
+				if (raf.getFilePointer() >= stoppoint) {
+					//since we subtract one for null byte below
+					raf.seek(raf.getFilePointer() + 1);
+					keepgoing = false;
+				}
+			}
+		}
+		long endpoint = raf.getFilePointer() - 1;
+		raf.seek(curpoint);
+		
+		byte[] byt = new byte[(int)(endpoint - curpoint)];
+		raf.read(byt);
+		try {
+			value = new String(byt, textEncoding);
+		} catch (Exception err) {
+			err.printStackTrace();
+			throw new BadValueException();
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 67:72, 147:152.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.